velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -97,60 +352,138 @@ RowVectorPtr NestedLoopJoinProbe::getOutput() {
-      childVector = BaseVector::wrapInDictionary(
-          BufferPtr(nullptr), buildIndices, numOutputRows, childVector);
-    }
-    output->childAt(projection.outputChannel) = childVector;
+bool NestedLoopJoinProbe::advanceProbeRows(vector_size_t probeCnt) {
+  probeRow_ += probeCnt;
+  if (probeRow_ < input_->size()) {
+    return false;
   }
-
-  probeRow_ += numProbeRows;
-  if (probeRow_ == inputSize) {
-    probeRow_ = 0;
+  probeRow_ = 0;
+  numPrevProbedRows_ = 0;
+  do {
     ++buildIndex_;
-    if (buildIndex_ == buildVectors_->size()) {
-      buildIndex_ = 0;
-      input_.reset();
+  } while (!hasProbedAllBuildData() &&
+           !buildVectors_.value()[buildIndex_]->size());
+  return hasProbedAllBuildData();
+}
+
+RowVectorPtr NestedLoopJoinProbe::doMatch(vector_size_t probeCnt) {
+  VELOX_CHECK_NOT_NULL(input_);
+  VELOX_CHECK(!hasProbedAllBuildData());
+
+  if (joinCondition_ == nullptr) {
+    return getCrossProduct(
+        probeCnt, outputType_, identityProjections_, buildProjections_);
+  }
+
+  auto filterInput = getCrossProduct(
+      probeCnt,
+      filterInputType_,
+      filterProbeProjections_,
+      filterBuildProjections_);
+
+  if (filterInputRows_.size() != filterInput->size()) {
+    filterInputRows_.resizeFill(filterInput->size(), true);
+  }
+  VELOX_CHECK(filterInputRows_.isAllSelected());
+
+  std::vector<VectorPtr> filterResult;
+  EvalCtx evalCtx(
+      operatorCtx_->execCtx(), joinCondition_.get(), filterInput.get());
+  joinCondition_->eval(0, 1, true, filterInputRows_, evalCtx, filterResult);
+  DecodedVector decodedFilterResult;
+  decodedFilterResult.decode(*filterResult[0], filterInputRows_);
+
+  const vector_size_t maxOutputRows_ = decodedFilterResult.size();
+  auto rawProbeOutMapping =
+      initializeRowNumberMapping(probeOutMapping_, maxOutputRows_, pool());
+  auto rawBuildOutMapping =
+      initializeRowNumberMapping(buildOutMapping_, maxOutputRows_, pool());
+  auto probeIndices = probeIndices_->asMutable<vector_size_t>();
+  auto buildIndices = buildIndices_->asMutable<vector_size_t>();
+  int32_t numOutputRows{0};
+  for (auto i = 0; i < maxOutputRows_; ++i) {
+    if (!decodedFilterResult.isNullAt(i) &&
+        decodedFilterResult.valueAt<bool>(i)) {
+      rawProbeOutMapping[numOutputRows] = probeIndices[i];
+      rawBuildOutMapping[numOutputRows] = buildIndices[i];
+      ++numOutputRows;
+    }
+  }
+  if (needsProbeMismatch(joinType_)) {
+    for (auto i = 0; i < numOutputRows; ++i) {
+      probeMatched_.setValid(rawProbeOutMapping[i], true);
+    }
+    probeMatched_.updateBounds();
+  }
+  if (needsBuildMismatch(joinType_)) {
+    for (auto i = 0; i < numOutputRows; ++i) {
+      buildMatched_[buildIndex_].setValid(rawBuildOutMapping[i], true);
     }
   }
-  return output;
-}
 
-bool NestedLoopJoinProbe::isFinished() {
-  return buildSideEmpty_ || (noMoreInput_ && input_ == nullptr);
+  if (numOutputRows == 0) {
+    return nullptr;
+  }
+  auto output =
```

Comment:
```
simply as?
return BaseVector::create<RowVector>();
```

Reply:
Seems there is no `create` function that takes no parameter.